sw/host/opentitanlib/src/test_utils/load_sram_program.rs

```diff
@@ -201,109 +203,86 @@ pub fn load_elf_sram_program(
     // Also note that there is a 6-byte gap between the end of .text and the beginning of .rodata
     // because .rodata needs a bigger alignment.
     //
-    // Instead, we elect to load segments that contain loadable data (and not BSS which is cleared
-    // by the SRAM CRT). Due to the gap between sections, we have to be careful because the CRT will
-    // compute the CRC of the entire data which encompasses all loadable sections but also the gaps
-    // between them! The code below explicitly writes the content of the gaps with known values that
-    // are taken into account in the CRC value.
+    // Demand paging has no use in embedded environment, and as shown above, if we load the
+    // program using the segments we could overwrite data unintentionally. Furthermore there will
+    // be an inconsistency between data loaded this way and data loaded via VMEM because the gap
+    // at the beginning is ignored by objcopy and is not covered by the CRC.
     //
-    // We also verify (read back and compare) the data from the section that contains the entry point.
+    // Fortunately there is a flag, confusingly named as `nmagic`, that changes the behaviour and
+    // disables this excessive alignment. The code below has a sanity check to ensure that the
+    // program is indeed compiled with `nmagic` enabeld by making sure tha the alignment does not
+    // exceed 8.
+    let crc = Crc::<u32>::new(&crc::CRC_32_ISO_HDLC);
+    let mut digest = crc.digest();
+    let mut last_address: Option<u32> = None;
+    for segment in file.segments() {
+        let address = segment.address();
+        let data = segment.data()?;
+
+        if data.is_empty() {
+            continue;
+        }
+
+        // It is much faster to load data word by word instead of bytes by bytes.
+        // The linker script always ensures that we the address and size are multiple of 4.
+        const WORD_SIZE: usize = std::mem::size_of::<u32>();
+        if address % WORD_SIZE as u64 != 0 || data.len() % WORD_SIZE != 0 {
+            bail!("The SRAM program must only consists of segments whose address and size are a multiple of the word size");
+        }
```

Comment:
These `if <cond> { bail!(...) }` statements can written with the `ensure!` macro:
```
ensure!(address % WORD_SIZE as u64 == 0 && data.len() % WORD_SIZE == 0, <error message>);
```

I'd also prefer to see structured errors rather than strings as errors.  This file already creates a structured error enumeration as `LoadSramProgramError`.  It should be pretty easy to add these error conditions to that enum.

Reply:
I don't think `LoadSramProgramError` is currently being used at all? The load function certainly only returns `anyhow::Result`.

Reply:
As a general practice on error results, I prefer structured errors.  The `anyhow::Result` type is a shorthand for `Result<T, anyhow::Error>`, and anyhow's error type can convert all implementors of the standard `Error` trait to itself (which is what `thiserror::Error` gives you: an easy way to create errors that implement `std::error::Error`).

The anyhow `bail`/`ensure` macros (IIRC) will perform the `Into` conversion for you, so you can simply write something like:
```
ensure!(condition_that_must_be_true, LoadSramProgramError::FooError("something bad happened"));
```

Occasionally, you'll have to explicitly specify `.into()`; usually when you're returning an `Err(...)` directly:
```
if (something_bad) {
    return Err(LoadSramProgramError::FooError("something bad").into());
}
```

We're using `anyhow::Result<T>` because it's "converts all errors" property makes returning errors up the call stack easy.  The `anyhow::Error` type supports downcasting back to the original type, so if we have a situation where an error needs to be inspected, we can downcast back to the original type an inspect the error.  Returning string errors rather than structured errors means error inspection turns into string parsing, which is generally not a best practice when dealing with errors.

Reply:
Okay, I see your point now. I thought you are asking me to change the return type to `Result<SramProgramInfo, LoadSramProgramError>`, which requires touching a lot more existing code. If you just want the downcastability then I can make the change.